packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/sync/SyncSession.kt

```diff
@@ -137,6 +142,17 @@ public interface SyncSession {
         progressMode: ProgressMode,
     ): Flow<Progress>
 
+    /**
+     * Create a [Flow] of _connection state changes_-events.
+     *
+     * Each change of state in the underlying connection will be emitted as a pair of the old and
+     * new [ConnectionState].
+     *
+     * The flow has an internal buffer of [Channel.BUFFERED] but if the consumer fails to consume
+     * the elements in a timely manner the flow will be completed with an [IllegalStateException].
+     */
+    public fun connectionState(): Flow<Pair<ConnectionState, ConnectionState>>
```

Comment:
I would turn this into something like `data class ConnectionStateChange(val oldState: ConnectionState, val newState: ConnectionState)` for better readability and consumption. Pairs are a bit annoying to reason when exposed in 3rd party API's.